core/macros/src/class.rs

```diff
@@ -0,0 +1,824 @@
+        // not a method and doesn't contain `self`.
+        let is_static =
+            take_path_attr(&mut fn_.attrs, "static") || !(has_explicit_method || has_receiver > 0);
+
+        Ok(Self {
+            length,
+            name,
+            body: quote! {
+                |   this: &boa_engine::JsValue,
+                    args: &[boa_engine::JsValue],
+                    context: &mut boa_engine::Context
+                | -> boa_engine::JsResult<boa_engine::JsValue> {
+                    let rest = args;
+                    #(#args_decl)*
+
+                    let result = Self:: #fn_name ( #(#args_call),* );
+                    boa_engine::TryIntoJsResult::try_into_js_result(result, context)
+                }
+            },
+            is_static,
+        })
+    }
+
+    fn getter(name: String, fn_: &mut ImplItemFn, class_ty: &Type) -> SpannedResult<Self> {
+        Self::method(name, true, fn_, class_ty)
+    }
+
+    fn setter(name: String, fn_: &mut ImplItemFn, class_ty: &Type) -> SpannedResult<Self> {
+        Self::method(name, true, fn_, class_ty)
+    }
+
+    fn constructor(fn_: &mut ImplItemFn, _class_ty: &Type) -> SpannedResult<Self> {
+        if fn_.sig.asyncness.is_some() {
+            error(&fn_.sig.asyncness, "Async methods are not supported.")?;
+        }
+
+        if !fn_.sig.generics.params.is_empty() {
+            error(&fn_.sig.generics, "Generic methods are not supported.")?;
+        }
+
+        let (args_decl, args_call): (Vec<TokenStream2>, Vec<TokenStream2>) = fn_
+            .sig
+            .inputs
+            .iter_mut()
+            .enumerate()
+            .map(|(i, a)| match a {
+                FnArg::Receiver(receiver) => error(receiver, "Constructors cannot use 'self'"),
+                FnArg::Typed(ty) => {
+                    let (_, decl, call) = Self::arg_from_pat_type(ty, i)?;
+                    Ok((decl, call))
+                }
+            })
+            .collect::<SpannedResult<_>>()?;
+
+        let length = take_length_from_attrs(&mut fn_.attrs)?.unwrap_or(args_decl.len());
+        let fn_name = &fn_.sig.ident;
+
+        // Does the function return Result<_> or JsResult<_>? If so, Into JsResult (to
+        // transform the error. If not, return Ok(_).
+        let return_statement = match &fn_.sig.output {
+            ReturnType::Default => quote! { Default::default() },
+            ReturnType::Type(_, ty) => {
+                if let Type::Path(path) = ty.as_ref() {
+                    let Some(t) = path.path.segments.last() else {
+                        return error(&fn_.sig.output, "Cannot infer return type.");
+                    };
+                    if t.ident == "Self" {
+                        quote! { Ok(result) }
+                    } else if t.ident == "JsResult" {
+                        quote! { result.into() }
+                    } else {
+                        return error(&fn_.sig.output, "Invalid return type.");
```

Comment:
Remark: Maybe we could return a more rusty error, noting/explaining what types are acceptable.